### PR TITLE
Refactor parts of Irrlicht's material system

### LINKED ISSUE #13621: Use bilinear filter when downscaling GUI images
## Problem

> Minetest does a horrible job at downscaling the larger version of the image.
> scrot
> I've tried other sizes, many were even worse. It looks like gui images are not even drawn with bilinear filter (tested with a different texture).

*Originally posted by @Desour in https://github.com/minetest/minetest/issues/13606#issuecomment-1597509092*

Minetest cannot upscale GUI images with bilinear filtering because that would blur pixel art, but we can *downscale* images with bilinear filtering. This would make downscaled images look much better. Compare these screenshots:

The Minetest logo on the about tab. The original resolution of the logo is 256x256 pixels. Downscaled without bilinear filtering.
screenshot before

The Minetest logo on the about tab. The original resolution of the logo is 256x256 pixels. Downscaled with bilinear filtering.
screenshot after

## Solutions

Use bilinear filtering when rendering 2D graphics, but just for downscaling textures, not for upscaling. This will require changes to IrrlichtMt, as it currently doesn't allow setting the minification and the magnification filter separately.

## Alternatives

None.

### PR DESCRIPTION
**Pull request by grorp**
This is the Minetest side of minetest/irrlicht#204 ("Refactor parts of the material system and set min filter to bilinear for 2D graphics"). The actual change is in the Irrlicht PR, this is "just" refactoring I did for that change.

Together with minetest/irrlicht#204, this PR fixes #13621. Because it exposes all OpenGL texture filtering options to Minetest, it also makes #13108 trivial to fix.

See minetest/irrlicht#204 for more information.

## To do

This PR is a Ready for Review.

## How to test

Use Minetest on all platforms and with all Irrlicht video drivers and verify that everything looks the same as before, with one exception: Images rendered with a lower resolution than their original resolution in GUIs look better. The difference is very visible with #13606, for example.

### CODE REVIEW

**Comment by grorp** [comment]
I see a problem here. If the Irrlicht PR is merged first, all Minetest builds will fail until this one is merged too. If this PR is merged first, all Minetest builds will fail until the Irrlicht PR is merged too.

As the changes affect many lines of code, I don't think using #defines in Minetest to make Minetest compatible with both Irrlicht versions is possible. The only solution is see is to merge both PRs and bump the IrrlichtMt tag at once. Does anyone have a better idea? Is this even possible?

**Comment by sfan5** [comment]
> If the Irrlicht PR is merged first, all Minetest builds will fail until this one is merged too.

They won't as the CI builds use a fixed tag from IrrlichtMt. This will only be an issue for users cloning both repos from the latest state.
this is also why the following suggestion exists: #13215

> The only solution is see is to merge both PRs and bump the IrrlichtMt tag at once.

Given the current situation this is still the best option.

**Comment by grorp** [comment]
Just to be clear, all this refactoring wasn't strictly required for separating the min and mag filters and setting the min filter to bilinear for 2D graphics.

I could have done this with a lot less changes, but that would have resulted in an inconsistent material API. I made all these changes because I wanted a consistent API for setting material properties.

Before this PR:
- `SMaterialLayer` has two boolean fields, `BilinearFilter` and `TrilinearFilter`, which select both the min and the mag filter
- you set them via `SMaterial::setFlag`, which automatically sets them on all texture layers for you

After this PR:
- `SMaterialLayer` has two enum fields, `MinFilter` (nearest, bilinear or trilinear) and `MagFilter` (nearest or bilinear)
- ❌ setting them via `SMaterial::setFlag`? impossible because they are no booleans
- ❌ manually setting them on all texture layers? would cause lots of code duplication
- ✅ you set them in a lambda passed to the new `SMaterial::forEachTexture` function

However, setting a per-texture material flag on all texture layers was the only real use of `SMaterial::setFlag`. All other uses could trivially be replaced by something like `material.<name of the flag> = <value>`, i.e. `setFlag` was a simple setter function. Therefore, I replaced all of its uses and removed it.

**Review by Desour — APPROVED** [review]
Code looks fine. LGTM.

File: src/client/clientmap.cpp [hunk 1]
```diff
@@ -842,32 +842,29 @@ void ClientMap::renderMap(video::IVideoDriver* driver, s32 pass)
...
 			// Apply filter settings
-			material.setFlag(video::EMF_TRILINEAR_FILTER,
-				m_cache_trilinear_filter);
-			material.setFlag(video::EMF_BILINEAR_FILTER,
-				m_cache_bilinear_filter);
-			material.setFlag(video::EMF_ANISOTROPIC_FILTER,
-				m_cache_anistropic_filter);
-			material.setFlag(video::EMF_WIREFRAME,
-				m_control.show_wireframe);
+			material.forEachTexture([this] (video::SMaterialLayer &tex) {
```
Comment (sfan5) [inline src/client/clientmap.cpp:845]: consider using `auto` here

**Review by sfan5 — APPROVED** [review]
tested together with #13606, works

File: src/client/clientmap.cpp (hunk 1 above)
Comment (grorp) [reply src/client/clientmap.cpp]: Now using `auto` where applicable.

**Comment by grorp** [comment]
Rebased, removed "(Minetest side)" from the commit descriptions, squashed a bit.

**Pull request merged**